butane_core/src/migrations/adb.rs

```diff
@@ -388,32 +449,113 @@ impl AColumn {
+                } else if let TypeKey::PK(referred_type) = referred_type_key {
+                    referred_table_name = referred_type.to_owned();
+                } else {
+                    unreachable!("Unexpected reference {:?}", self.reference);
+                }
+                if let Some(table) = tables.get(&referred_table_name) {
+                    if let Some(pk) = table.pk() {
+                        self.reference = Some(ARef::Literal(ARefLiteral::new(
+                            referred_table_name,
+                            pk.name.clone(),
+                        )));
+                        true
+                    } else {
+                        false
+                    }
+                } else {
+                    false
+                }
+            }
+            _ => unreachable!("can only resolve deferred references"),
+        }
+    }
+
     pub fn is_auto(&self) -> bool {
         self.auto
     }
 }
 
+/// Create a table for the [crate::many::Many] relationship.
+/// Should not be used directly, except in tests.
+pub fn create_many_table(
+    main_table_name: &str,
+    many_field_name: &str,
+    many_field_type: &DeferredSqlType,
+    main_table_pk_field_name: &str,
+    main_table_pk_field_type: &DeferredSqlType,
+) -> ATable {
+    let mut table = ATable::new(format!("{main_table_name}_{many_field_name}{MANY_SUFFIX}"));
+    let col = AColumn::new(
+        "owner",
+        main_table_pk_field_type.clone(),
+        false, // nullable
+        false, // pk
+        false, // auto
+        false, // unique
+        None,  // default
+        Some(ARef::Literal(ARefLiteral::new(
+            main_table_name.to_owned(),
+            main_table_pk_field_name,
+        ))),
+    );
+    table.add_column(col);
+    let mut col = AColumn::new_simple("has", many_field_type.clone());
+    if matches!(many_field_type, DeferredSqlType::Deferred(TypeKey::PK(_))) {
+        col.add_reference(&ARef::Deferred(many_field_type.clone()));
+    }
+    table.add_column(col);
+    table
+}
+
 /// Individual operation use to apply a migration.
 #[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
 pub enum Operation {
     //future improvement: support column renames
+    /// Add a table.
     AddTable(ATable),
+    /// Add table constraints referring to other tables, if the backend supports it.
+    AddTableConstraints(ATable),
+    /// Add a table, if it doesnt already exist.
     AddTableIfNotExists(ATable),
+    /// Remove named table.
     RemoveTable(String),
+    /// Add a table column.
     AddColumn(String, AColumn),
+    /// Remove a table column.
     RemoveColumn(String, String),
+    /// Change a table columns type.
     ChangeColumn(String, AColumn, AColumn),
 }
 
 /// Determine the operations necessary to move the database schema from `old` to `new`.
 pub fn diff(old: &ADB, new: &ADB) -> Vec<Operation> {
+    // TODO: re-order tables so tables without references go first,
```

Comment:
need to create a test, effectively same as existing many tests but with a & b swapped to prove this is a problem, and then do the fix